File: Minimum_Path_Sum.cpp
class Solution {
public:
    /**
     * @param grid: a list of lists of integers.
     * @return: An integer, minimizes the sum of all numbers along its path
     */
    int minPathSum(vector<vector<int> > &grid) {
        // write your code here
        int rows = grid.size();
        int cols = grid[0].size();
        vector<vector<int>> table(2, vector<int>(cols, 0));
        table[0][0] = grid[0][0];
        for(int i = 1; i < cols; ++i)
            table[0][i] = grid[0][i] + table[0][i-1];
        for(int i = 1; i < rows; ++i){
            table[i%2][0] = table[(i-1)%2][0] + grid[i][0];
            for(int j = 1; j < cols; ++j){
                table[i%2][j] = min(table[(i-1)%2][j], table[i%2][j-1]) + grid[i][j];
            }
        }
        return table[(rows-1)%2][cols-1];
    }
};

File: Minimum Path Sum.py
class Solution:
    """
    @param grid: a list of lists of integers.
    @return: An integer, minimizes the sum of all numbers along its path
    """
    def minPathSum(self, grid):
        # write your code here
        
        m, n = len(grid), len(grid[0])
        ret = [[0] * (n) for __ in range(m)]
        ret[0][0] = grid[0][0]
        for i in range(1, n):
            ret[0][i] = grid[0][i] + ret[0][i - 1]
        for i in range(1, m):
            ret[i][0] = ret[i - 1][0] + grid[i][0]
            for j in range(1, n):
                ret[i][j] = min(ret[i][j - 1], ret[i - 1][j]) + grid[i][j]
        return ret[m - 1][n - 1]